### PR TITLE
Connect dashboard to API

### PR DESCRIPTION
**Pull request by plausibly**
- Periodically fetch post data from the API and load it onto dashboard
- "Read more" does another API request which fetches the rest of the post data
- Delete button is functional
- Added UserContext which contains the current logged in user
- Dashboard forces you to login if you are not properly authorized
- Sign Out button integrated with API
- Delete/Edit only show up if the user is authorized
- Style / Bug fixes

Notes: Need to update POST type after Daniel's PR is merged.

### CODE REVIEW

**Comment by andy9775** [comment]
I'm confused as to why this was reopened.

**Comment by plausibly** [comment]
> I'm confused as to why this was reopened.

I was updating some things. We did not merge it to main branch

File: client/src/components/CreatePost.tsx [hunk 1]
```diff
@@ -37,14 +53,31 @@ function CreatePost() {
...
   return (
+    <>
+      <Button
+        variant="outlined"
+        onClick={() => {
+          toggleDialog(true);
+        }}
+        data-testid="btn-newpost"
+      >
+        New Post
```
Comment (andy9775) [inline client/src/components/CreatePost.tsx:71]: `Create a Post`?

File: client/src/containers/PostDashboard.tsx [hunk 2]
```diff
@@ -1,13 +1,49 @@
...
+      })
+      .then((res: any) => {
+        if (res.data.data.result) {
+          updateRecent(res.data.data.result);
+        }
+      })
+      .catch((err) => console.log(err));
+  };
+
+  React.useEffect(() => {
```
Comment (andy9775) [inline client/src/containers/PostDashboard.tsx:31]: We could store the posts in app wide state, set the polling logic in app.tsx, then this way we will always be polling not just when the user is on the posts dashboard.

**Review by andy9775 — CHANGES_REQUESTED** [review]
I noticed some inconsistencies with double/single quotes. Let's make sure we stick to one. A note on deleting posts (and things in general limited to a specific user). Hit the /me endpoint to get user info, then based on the uid of the post and the uid of the current user, we can enable/disable certain things. We should use something like redux or context to store the state of the current user so we only have to fetch it once. 

Otherwise, request a review once you update the client api stuff.

File: client/src/containers/PostDashboard.tsx (hunk 2 above)
Comment (plausibly) [reply client/src/containers/PostDashboard.tsx:31]: The site does always poll. Since "View post" and "New post" are in dialogs, post dashboard is technically open in the background.

File: client/src/components/CreatePost.tsx (hunk 1 above)
Comment (plausibly) [reply client/src/components/CreatePost.tsx:71]: I'll actually revert this file to the state of GDSC main since Gurvir has this in his PR

**Comment by danielduong8** [comment]
comment on my PR so that i ermember to integrate sign out

**Comment by plausibly** [comment]
I updated the PR message at the top to reflect my changes

File: client/src/containers/PostDashboard.tsx [hunk 3]
```diff
@@ -46,14 +57,46 @@ function RecentPosts() {
...
 export default function PostDashboard() {
+  const [currentUser, setUser] = React.useState(null);
+  const [isLoading, setLoading] = React.useState(true);
+  const navigate = useNavigate();
+
+  const authBarrier = () => {
+    // force a login
+    navigate('/');
+    setUser(null);
+  };
```
Comment (andy9775) [inline client/src/containers/PostDashboard.tsx:68]: We could to something like this using react router, this way it automatically does checks on all pages.

File: client/src/components/ViewPostDialog.tsx [hunk 4]
```diff
@@ -183,26 +201,39 @@ export default function ViewPostDialog(props: {
...
-        if (res.data) {
-          setData(res.data.data.result);
+    api
+      .fetchPost(props.postUser.id)
+      .then(
+        (
+          res: APIResponse<{ data: { result?: PostUser }; message?: string }>
+        ) => {
+          if (res.data && res.data.data && res.data.data.result) {
+            // appeasing TS
```
Comment (andy9775) [inline client/src/components/ViewPostDialog.tsx:216]:
```suggestion

```
Doesn't really add value

**Review by andy9775 — CHANGES_REQUESTED** [review]
Some first pass comments.

Can you undo the formatting changes for now? And only have them in the formatting PR? You can do this by having save on format, format only the changed lines and not the whole file.

I would also add a test which checks for the post being created. Mock the `api` class and assert that posts are created when the data is valid. Then, mock the return and make sure it's being displayed correctly. Same for pagination, mock a list of fake data being returned and make sure that the first and last elements for  example are being updated.

File: client/src/containers/PostDashboard.tsx [hunk 5]
```diff
@@ -5,26 +5,37 @@ import Container from '@mui/material/Container';
...
-
-import { api_v1 } from '../api/v1';
 import CreatePost from '../components/CreatePost';
 
+import ServerApi, { APIResponse, PostUser } from '../api/v1';
+import { useNavigate } from 'react-router-dom';
+
+const api = new ServerApi();
+
+export const UserContext: React.Context<{
```
[the remark this answers is not in the document]
Comment (plausibly) [reply client/src/containers/PostDashboard.tsx:19]: But the user will not be authenticated on App.tsx, we only want to get the user after authentication (which will route them to the dashboard). The rest of our site is contained on the dashboard

File: client/src/api/v1/index.ts [hunk 6]
```diff
@@ -1,4 +1,101 @@
-import axios from 'axios';
+import axios, { AxiosInstance } from 'axios';
+import { Post } from 'models/post';
+import { User } from 'models/user';
 
-export const api_v1 = axios.create({ baseURL: `api/v1` });
-// TODO: This will be changed to Daniel's API class
+export type PostUser = Post & {
```
[the remark this answers is not in the document]
Comment (plausibly) [reply client/src/api/v1/index.ts:5]: I think this is fine for now, we dont want to make the controller class publicly available.

**Comment by plausibly** [comment]
I wasn't able to fix the bug where the currently viewed post would change if a new post was made in the background. Instead, when a user is viewing a post the dashboard will not poll data. I also updated useEffect to fetch instantly once they close out of the post, and then resume polling.

File: client/src/containers/PostDashboard.tsx (hunk 3 above)
Comment (plausibly) [reply client/src/containers/PostDashboard.tsx:93]: How do you mean? Like disabling the "/dashboard" route on App.tsx until they are authed?

File: client/src/components/PostPreview.tsx [hunk 7]
```diff
@@ -10,13 +10,45 @@ import Typography from "@mui/material/Typography";
...
 
+  /**
+   * Return a string of the date relative to now in minutes, hours or days. If the date
+   * is over a month, return the date stringified. 
+   * 
+   * @param base - The base date being compared with
+   * @param date - The date converted into relative time
+   * @returns String containing the date relative to now
+   */
+function relativeTime(date: Date) {
```
Comment (andy9775) [inline client/src/components/PostPreview.tsx:23]: There should be libraries which do this for us. Benefit would be that all cases are handled and they are tested.

File: client/src/components/ViewPostDialog.tsx [hunk 8]
```diff
@@ -70,11 +69,23 @@ function MoreOptions(props: {
...
+          setMsg('Post has been reported.');
+        } else {
+          setMsg('Failed to report post. ');
+        }
+        showAlert(true);
+      })
+      .catch(() => {
+        setMsg('Failed to report post.');
+        showAlert(true);
+      });
```
Comment (andy9775) [inline client/src/components/ViewPostDialog.tsx:87]:
```suggestion
    api
      .reportPost(props.postID)
      .then((res) => {
        if (res.status === 204) {
          setMsg('Post has been reported.');
        } else {
          setMsg('Failed to report post. ');
        }
      })
      .catch(() => {
        setMsg('Failed to report post.');
      }).finally(() => showAlert(true));
```

File: client/src/containers/PostDashboard.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (andy9775) [reply client/src/containers/PostDashboard.tsx:21]: It could be an empty object. The typical pattern is to have user data in something like redux which is app wide state.

I feel like there would be a need to make a change in the future, but I'll leave it up to you.

File: client/src/containers/PostDashboard.tsx [hunk 9]
```diff
@@ -11,46 +11,68 @@ import Header from '../components/Header';
 import PostPreview from '../components/PostPreview';
 import CreatePost from '../components/CreatePost';
 
-import ServerApi, { APIResponse, PostUser } from '../api/v1';
+import ServerApi, { PostUserPreview } from '../api/v1';
 import { useNavigate } from 'react-router-dom';
```
Comment (andy9775) [inline client/src/containers/PostDashboard.tsx]:
```suggestion
import { useNavigate } from 'react-router-dom';

import ServerApi, { PostUserPreview } from '../api/v1';
```

File: client/src/containers/PostDashboard.tsx [hunk 10]
```diff
@@ -11,46 +11,68 @@ import Header from '../components/Header';
...
 
+  /* Fetch posts triggered by page-change or post dialog close */
+  React.useEffect(() => {
+    if (!openedPost) {
+      api
+        .fetchRecentPosts(POSTS_PER_PAGE, POSTS_PER_PAGE * (props.pageNum - 1))
+        .then((res) => {
+          if (res.data && res.data.data.result) {
+            updateRecent(res.data.data.result);
+            props.setPageCount(Math.ceil(res.data.data.count / POSTS_PER_PAGE));
```
Comment (andy9775) [inline client/src/containers/PostDashboard.tsx:65]: I would write some tests for the pagination logic to make sure the correct page is being displayed and the correct data is being displayed. You can just create some mock data to be returned from the mock `fetchRecentPosts` method and just assert that the UI displays correctly.

File: client/src/containers/PostDashboard.tsx [hunk 11]
```diff
@@ -141,7 +174,7 @@ export default function PostDashboard() {
             color='text.secondary'
             component='p'
           >
-            You're all up to date!
+            You"re all up to date!
```
Comment (andy9775) [inline client/src/containers/PostDashboard.tsx:177]:
```suggestion
            You're all up to date!
```

File: client/src/containers/PostDashboard.tsx [hunk 12]
```diff
@@ -125,7 +150,15 @@ export default function PostDashboard() {
             alignItems: 'center',
           }}
         >
-          <Pagination count={1} color='primary' variant='outlined' />
+          <Pagination
+            count={pageCount}
+            page={page}
+            onChange={(event: React.ChangeEvent<unknown>, pg: number) => {
```
Comment (andy9775) [inline client/src/containers/PostDashboard.tsx:156]:
```suggestion
            onChange={(event: React.ChangeEvent<{}>, pg: number) => {
```

File: server/controllers/v1/post.ts [hunk 13]
```diff
@@ -26,25 +36,34 @@ export default class PostController {
...
 
     return {
-      status: data.count > 0 ? 200 : 204,
-      data: { result: data.rows as PostUser[], count: data.count },
+      status: data[0].count > 0 ? 200 : 204,
+      data: {
+        result: data[0].rows as any as PostUserPreview[],
+        count: data[0].count,
+        total: data[1],
+      },
```
Comment (andy9775) [inline server/controllers/v1/post.ts:66]: Can you write a test for this to ensure that the data is correct?

**Review by andy9775 — CHANGES_REQUESTED** [review]
If deciding to keep the `relativeTime` function, make sure you have proper tests, it seems edge case-y.

I would also add some test for the pagination stuff. Next review should primarily be that.

File: client/src/containers/PostDashboard.tsx (hunk 12 above)
Comment (plausibly) [reply client/src/containers/PostDashboard.tsx:156]: I have to explicitly put unknown or it causes an error, I looked into this on the MUI docs and they also put unknown for this.

``Type 'ChangeEvent<unknown>' is not assignable to type 'ChangeEvent<{}>'.``

File: server/controllers/v1/post.ts (hunk 13 above)
Comment (plausibly) [reply server/controllers/v1/post.ts:66]: sure

File: client/src/components/PostPreview.tsx (hunk 7 above)
Comment (plausibly) [reply client/src/components/PostPreview.tsx:22]: I added some tests for this

File: client/src/api/v1/index.ts [hunk 14]
```diff
@@ -13,10 +13,10 @@ export type PostUserPreview = {
   title: string;
   createdAt: string;
 } & {
-  User: { id: string; firstName: string; lastName: string; userName: string };
+  User: { id: string; firstName: string; lastName: string };
 };
 
-export class ApiError extends Error {}
+class ApiError extends Error {}
```
Comment (andy9775) [inline client/src/api/v1/index.ts:19]: This error should be caught and handled.

File: client/src/containers/PostDashboard.test.tsx [hunk 15]
```diff
@@ -54,6 +55,196 @@ describe("Test interaction with dashboard", () => {
...
+      .mockImplementation((limit, offset) => {
+        const results = posts.slice(offset, limit + offset);
+        return Promise.resolve({
+          status: 200,
+          data: {
+            data: {
+              result: results,
+              count: results.length,
+              total: posts.length,
+              message: 'asd',
```
Comment (andy9775) [inline client/src/containers/PostDashboard.test.tsx:215]:
```suggestion
```
Message is optional, and is only returned if an error occurs.

**Review by andy9775 — APPROVED** [review]
LGTM!

File: client/src/containers/PostDashboard.tsx (hunk 3 above)
Comment (andy9775) [reply client/src/containers/PostDashboard.tsx:93]: https://ui.dev/react-router-v5-protected-routes-authentication/

On loading a route you can check if the user object exists in app state. If not, redirect out to the login page, otherwise display the component for that page. 

```typescript
export function ProtectedRoute(props: {component: React.Component}) {
    const [user, ...] = getUserFromContext(); // some logic here
   if (user & user.isAuthenticated) {
     return <component/>
  }
  redirectToLoginPage(); // you get it
}
```

It's much safer to have it centralized in one location

File: client/src/App.tsx [hunk 16]
```diff
@@ -1,33 +1,83 @@
...
-          {/* TODO: Ensure authenticated then redirect to dashboard */}
-        </Routes>
-      </BrowserRouter>
-    </ThemeProvider>
+      <ThemeProvider theme={theme}>
+        <CssBaseline />
+        <BrowserRouter>
+          <Routes>
+            <Route path='/' element={<AuthContainer setAuthed={setAuthed}/>} />
+            <Route path='/dashboard' element={<ProtectedRoute isAuth={authed}/>} />
```
Comment (danielduong8) [inline client/src/App.tsx:75]: I think in cases like these it's probably a good idea to use contexts rather than passing down a state hook.

**Review by danielduong8 — APPROVED** [review]
